### PR TITLE
Implement Stage 2

### PR DESCRIPTION
**Pull request by BeksOmega**
### Description

Depends on #10 

Implements all of the abstraction stuff:
* Sanitizing method and getter return values.
* Sanitizing event parameter values.
* Dropdown blocks returning different values depending on companion version.
* Allowing dropdown blocks to be dictionary keys.

Important Note: I couldn't get the protect-enum macro to work from the getFormYail() function, and I couldn't get *any* logging working. So I need some help :/

And another note: This PR is maximally backwards compatible. If you think any backwards compatibility stuff should be removed for maintainability, simplicity, etc, I am happy to remove it =)

### Testing

* Tested that methods and getters are sanitized properly depending on the version of the companion.
* Tested that event params are sanitized properly depending on the version of the companion.
* Tested that enum values can be used to access both the value and the underlying value from the dictionary.

Both this:
blocks (2)

And this:
blocks (3)

return "East value"!

### Recommended Method for Review
Commit-Wise

### CODE REVIEW

**Comment by BeksOmega** [comment]
Note to self: If this gets merged I need to update the docs to include info about OptionLists as parameter/return types.
- [ ] Docs updated.

**Comment by BeksOmega** [comment]
Discussed today: Mr. Lyn would like to break backwards compatibility wrt to comparing getters with two different abstract types, so that we get some of the benefits of this sanitization.

So in the past you could compare these two blocks and under certain circumstances it would be true because they both return ints:

blocks

We would now like this to always be false because they have different abstract types (HorizontalAlignment and VerticalAlignment).

This is a simple fix (I think it just involves reverting some changes I made in this PR) but I'm going to focus on getting screen blocks and asset blocks done first.

[EDIT: [Done](https://github.com/BeksOmega/appinventor-sources/pull/24/commits/844d364309f8118e804602f3620be7a3915229f6)]

File: appinventor/blocklyeditor/src/blocks/helpers.js [hunk 1]
```diff
@@ -0,0 +1,143 @@
+// -*- mode: java; c-basic-offset: 2; -*-
+// Copyright 2013-2014 MIT, All rights reserved
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/blocks/helpers.js]: Copyright year should be 2020 as this is a new file.

File: appinventor/components/src/com/google/appinventor/components/scripts/ComponentProcessor.java [hunk 2]
```diff
@@ -1746,9 +1746,7 @@ private HelperKey elementToHelperKey(Element elem, TypeMirror type) {
   private HelperKey hasOptionListHelper(Element elem, TypeMirror type) {
     // Check if the elem type is an OptionList
     if (isOptionList(type)) {
-      throw new RuntimeException("Using OptionLists as parameter types is not yet supported. Please " +
-          "use a concrete value (such as int, or String) annotated with an @Options parameter.");
-      //return optionListToHelperKey(((DeclaredType)type).asElement());
+      return optionListToHelperKey(((DeclaredType)type).asElement());
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/scripts/ComponentProcessor.java]: Minor whitespace issue: space between `(DeclaredType)` and `type`.

File: appinventor/components/src/com/google/appinventor/components/runtime/OptionHelper.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        }
+        SimpleProperty prop = m.getAnnotation(SimpleProperty.class);
+        if (prop != null) {
+          methodMap.put(methodKey, m);
+        }
+      }
+    }
+    return methodMap;
+  }
+}
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/OptionHelper.java]: Minor whitespace issue: Missing newline at EOF

File: appinventor/components/src/com/google/appinventor/components/runtime/Form.java [hunk 4]
```diff
@@ -1640,7 +1640,7 @@ public void AlignVerticalAbstract(VerticalAlignment alignment) {
   @SimpleProperty(category = PropertyCategory.APPEARANCE,
     description = "The animation for switching to another screen. Valid" +
     " options are default, fade, zoom, slidehorizontal, slidevertical, and none"    )
-  public String OpenScreenAnimation() {
+  public @Options(ScreenAnimation.class) String OpenScreenAnimation() {
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/Form.java:1643]: Maybe I am missing something, but why is this change here instead of another PR related specifically to the ScreenAnimation? It seems like this could have been done in stage 1.

File: appinventor/components/src/com/google/appinventor/components/runtime/Form.java [hunk 5]
```diff
@@ -1692,7 +1692,7 @@ public void OpenScreenAnimationAbstract(ScreenAnimation animType) {
     description = "The animation for closing current screen and returning " +
     " to the previous screen. Valid options are default, fade, zoom, slidehorizontal, " +
     "slidevertical, and none")
-  public String CloseScreenAnimation() {
+  public @Options(ScreenAnimation.class) String CloseScreenAnimation() {
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/Form.java:1695]: Ditto.

File: appinventor/components/src/com/google/appinventor/components/runtime/OptionHelper.java [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+import com.google.appinventor.components.annotations.Options;
+import com.google.appinventor.components.annotations.SimpleEvent;
+import com.google.appinventor.components.annotations.SimpleFunction;
+import com.google.appinventor.components.annotations.SimpleProperty;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.HashMap;
+import java.util.Map;
+import android.util.Log;
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/OptionHelper.java]: Sort order of imports should have android come first.

File: appinventor/components/src/com/google/appinventor/components/runtime/OptionHelper.java [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+public class OptionHelper {
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/OptionHelper.java:24]: Add a Javadoc to describe the class.

File: appinventor/components/src/com/google/appinventor/components/runtime/OptionHelper.java [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+  public static final String TAG = "OptionHelper";
+
+  public static Map<String, Map<String, Method>> componentMethods =
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/OptionHelper.java]: I don't think we want this to be public, otherwise other classes (e.g., extensions) could modify it. Also, is there any reason it couldn't be final?

File: appinventor/components/src/com/google/appinventor/components/runtime/OptionHelper.java [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+      new HashMap<String, Map<String, Method>>();
+    
+  /**
+   * Returns the OptionList version of the value if the function's return type has an @Options
+   * annotation notating that the value can be coerced to an OptionList.
+   */
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/OptionHelper.java:37]: Add `@param` and `@returns` Javadoc annotations to describe the parameters/return value.

File: appinventor/components/src/com/google/appinventor/components/runtime/OptionHelper.java [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+    if (annotation == null) {
+      return value;
+    }
+    Class<?> optionListClass = annotation.value();
+    Object newValue = null;
+    try {
+      Method fromValue = optionListClass.getMethod("fromUnderlyingValue", value.getClass());
+      newValue = fromValue.invoke(optionListClass, value);
+    } finally {
+      if (newValue != null) {
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/OptionHelper.java]: It's not really standard to return from inside a finally block. You should catch the exception(s) thrown in the try block and return value from there as well as return from inside the try block.

File: appinventor/components/src/com/google/appinventor/components/runtime/OptionHelper.java [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+        return newValue;
+      }
+      return value;
+    }
+  }
+
+  /**
+   * Returns the args after any coercable args have been coerced to an OptionList. An arg is
+   * coercable if the function header has an @Options annotation associated with that arguement.
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/OptionHelper.java:70]: Add `@param` and `@returns` annotations to describe the arguments and return values.

File: appinventor/components/src/com/google/appinventor/components/runtime/OptionHelper.java [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+    Annotation[][] paramAnnotations = calledFunc.getParameterAnnotations();
+    int i = 0;
+    for (Annotation[] annotations : paramAnnotations) {
+      for (Annotation annotation : annotations) {
+        if (annotation.annotationType() == Options.class) {
+          Options castAnnotation = (Options) annotation;
+          Class<?> optionListClass = castAnnotation.value();
+          try {
+            Method fromValue = optionListClass.getMethod("fromUnderlyingValue", args[i].getClass());
+            args[i] = fromValue.invoke(optionListClass, args[i]);
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/OptionHelper.java]: Given that fromUnderlyingValue could return false, you will want to capture the return value and only replace the argument if it is null. Otherwise, a call to a method that takes an `int` will cause a null dereference trying to unwrap the null Integer object into its primitive value.

Alternatively, we may want to throw a RuntimeError here due to the arguments not being coercible (see coerce-args in runtime.scm).

File: appinventor/components/src/com/google/appinventor/components/runtime/OptionHelper.java [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+   * Returns a map populated with all relevant Methods of the given Class. This includes all events,
+   * property getters, and non-void methods.
+   */
+  private static Map<String, Method> populateMap(Class<?> clazz) {
+    Map<String, Method> methodMap = new HashMap<String, Method>();
+    Method[] methods = clazz.getMethods();
+
+    // Add all the relevant methods to the map.
+    for (Method m : methods) {
+      String methodKey = m.getName();
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/OptionHelper.java:149]: May want to check that we are only doing this for public methods.

File: appinventor/blocklyeditor/src/blocks/components.js [hunk 14]
```diff
@@ -1200,22 +1223,17 @@ Blockly.Blocks.component_set_get = {
...
-    if (inputOrOutput == Blockly.Blocks.Utilities.INPUT) {
-      var helperType = Blockly.Blocks.Utilities
-          .helperKeyToBlocklyType(property.helperKey, this);
-      if (helperType && helperType != blocklyType) {
-          check.push(helperType);
-      }
+    var helperType = Blockly.Blocks.Utilities
+        .helperKeyToBlocklyType(property.helperKey, this);
+    if (helperType && helperType != blocklyType) {
+        check.push(helperType);
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/blocks/components.js]: Indentation should only be 2 in from the if statement.

File: appinventor/buildserver/src/com/google/appinventor/buildserver/resources/runtime.scm [hunk 15]
```diff
@@ -742,6 +742,14 @@
        #`(begin
            (define-event-helper ,(gen-generic-event-name #`component-type #`event-name) args body))))))
 
+
+(define-syntax protect-enum
```
Comment (ewpatton) [inline appinventor/buildserver/src/com/google/appinventor/buildserver/resources/runtime.scm]: `protect-enum` actually needs to be sent as part of the companion YAIL from the browser as well as older companions won't have the macro definition.

File: appinventor/blocklyeditor/src/generators/yail/helpers.js [hunk 16]
```diff
@@ -12,20 +12,18 @@ Blockly.Yail['helpers_dropdown'] = function() {
...
+  // and if it does it will return the abstract enum value. If the companion
+  // does not support OptionLists it will continue to return the concrete value.
+  var code = '(protect-enum ' + enumValue + ' ' + concreteValue + ')';
 
-  // Currently we are returning the concrete values of the optionList option for
-  // easy backwards compatibility with the companion. But in the future we will
-  // return a macro which checks if the companion supports OptionLists and if
-  // it does it will return the abstract enum value. If the companion does not
-  // support OptionLists it will continue to return the concrete value.
+  return [code, Blockly.Yail.ORDER_ATOMIC]
```
Comment (ewpatton) [inline appinventor/blocklyeditor/src/generators/yail/helpers.js]: Missing semicolon

File: appinventor/buildserver/src/com/google/appinventor/buildserver/resources/runtime.scm [hunk 17]
```diff
@@ -1182,6 +1191,15 @@
    ((instance? data JavaCollection) (java-collection->yail-list data))
    (#t (sanitize-atomic data))))
 
+(define (sanitize-return-value component func-name value)
+  (define-alias OptionHelper com.google.appinventor.components.runtime.OptionHelper)
+  (if (enum? value)
+      value
```
Comment (ewpatton) [inline appinventor/buildserver/src/com/google/appinventor/buildserver/resources/runtime.scm]: Only 2 spaces indent from `(`

File: appinventor/buildserver/src/com/google/appinventor/buildserver/resources/runtime.scm [hunk 18]
```diff
@@ -1182,6 +1191,15 @@
...
+      (let ((value (OptionHelper:optionListFromValue component func-name value)))
+      (if (enum? value)
```
Comment (ewpatton) [inline appinventor/buildserver/src/com/google/appinventor/buildserver/resources/runtime.scm:1192]: if statement should be indented 2 from the start of the let statement.

File: appinventor/buildserver/src/com/google/appinventor/buildserver/resources/runtime.scm [hunk 19]
```diff
@@ -1182,6 +1191,15 @@
...
+          value
```
Comment (ewpatton) [inline appinventor/buildserver/src/com/google/appinventor/buildserver/resources/runtime.scm]: Same on indentation here.

File: appinventor/buildserver/src/com/google/appinventor/buildserver/resources/runtime.scm (hunk 15 above)
Comment (ewpatton) [reply appinventor/buildserver/src/com/google/appinventor/buildserver/resources/runtime.scm]: Might want to add a comment before to document why/what it is for.

File: appinventor/buildserver/src/com/google/appinventor/buildserver/resources/runtime.scm [hunk 20]
```diff
@@ -2570,8 +2589,11 @@ Dictionary implementation.
             (*:get (as YailDictionary yail-dictionary) key))
           (#t default))))
     (if (eq? result #!null)
-      default
-      result)))
+        ;; if we don't find anything associated with the abstract type, try the underlying type.
```
Comment (ewpatton) [inline appinventor/buildserver/src/com/google/appinventor/buildserver/resources/runtime.scm]: Indentation here should be 2 for each new scope.

**Review by ewpatton — CHANGES_REQUESTED** [review]
So one thing that is different with this is how the enum values are handled when coming from the block versus coming from a component. The enum still appears as the underlying value but the value returned from a component will show the enum value name. Personally, I prefer the latter but we should make a decision so it is consistent.

image

File: appinventor/components/src/com/google/appinventor/components/runtime/Form.java (hunk 4 above)
Comment (BeksOmega) [reply appinventor/components/src/com/google/appinventor/components/runtime/Form.java:1643]: Yeah these changes do belong in Stage 1, I just missed them. ScreenAnimation was one of the first 3 properties I upgraded, and we were originally thinking you would only need to add the `@Options` annotation on either the getter or the setter of a getter-setter pair (not both).

When we changed this over I forgot to upgrade it, but it still worked in Stage 1. But Stage 2 requires annotations on both, which is how I discovered this bug.

I can separate the change into a different PR if ya like.

File: appinventor/components/src/com/google/appinventor/components/runtime/OptionHelper.java (hunk 6 above)
Comment (BeksOmega) [reply appinventor/components/src/com/google/appinventor/components/runtime/OptionHelper.java]: Just removed it since this file no longer includes logging.

File: appinventor/components/src/com/google/appinventor/components/runtime/OptionHelper.java (hunk 8 above)
Comment (BeksOmega) [reply appinventor/components/src/com/google/appinventor/components/runtime/OptionHelper.java]: Changed to private static final =)

File: appinventor/components/src/com/google/appinventor/components/runtime/OptionHelper.java (hunk 12 above)
Comment (BeksOmega) [reply appinventor/components/src/com/google/appinventor/components/runtime/OptionHelper.java]: I'm a bit confused about what you're saying. Shouldn't fromUnderlyingValue always return a valid OptionList in this case because this function is for coercing args of events? For example:

``` java
@SimpleEvent
public void MyTestEvent(@Options(SomeEnum.class) int val) {
  dispatchEvent(this, "MyTestEvent", val);
}
```

In the above case shouldn't val always be a valid underlying value of SomeEnum? If there aren't any bugs in the component that is.

I kind of feel like that's not what you're talking about though. Could you give me an example of the problem you're thinking of?

File: appinventor/buildserver/src/com/google/appinventor/buildserver/resources/runtime.scm (hunk 15 above)
Comment (BeksOmega) [reply appinventor/buildserver/src/com/google/appinventor/buildserver/resources/runtime.scm]: I know but I couldn't get it working :/ I tried to add the macro via the getFormYail() function, but it wasn't working. I also couldn't get *any* logging to work, so I didn't know how to debug it.

Do you have any advice / know how I could add some logging?

**Comment by BeksOmega** [comment]
> The enum still appears as the underlying value but the value returned from a component will show the enum value name. Personally, I prefer the latter but we should make a decision so it is consistent.

This is only because I didn't bump the BLOCKS_LANGUAGE_VERSION, but you've loaded up a new companion so return values are getting sanitized. Basically you've got an inbetween-y companion. On old companions everything will be concrete. And on new companions everything will be abstract.

I've gone ahead and added a bump to the lang version to this PR.

**Comment by ewpatton** [comment]
@BeksOmega Thanks that worked for me.

File: appinventor/buildserver/src/com/google/appinventor/buildserver/resources/runtime.scm (hunk 15 above)
Comment (ewpatton) [reply appinventor/buildserver/src/com/google/appinventor/buildserver/resources/runtime.scm]: There is the `(android-log)` function that you can use. There are other examples in this file where it is used.

File: appinventor/components/src/com/google/appinventor/components/runtime/Form.java (hunk 4 above)
Comment (ewpatton) [reply appinventor/components/src/com/google/appinventor/components/runtime/Form.java:1643]: Assuming this gets merged it won't matter much because we will squash everything into a single change.

**Comment by BeksOmega** [comment]
@ewpatton Is getFormYail the right place to add the macro? I did the following steps:

1) Load local AI in browser.
2) Set a breakpoint in the ode-0.js file inside the getFormYail function.
3) Connected my local instance to the companion.
4) Ran "do it" on one of my dropdown blocks.

And my breakpoint never got hit. It seems like getFormYail no longer gets called? I'm going to keep investigating, but I figured I'd give you a ping on it.

**Comment by ewpatton** [comment]
Hmm. That’s right. getFormYail is called to save the YAIL to the server, which will never need the macro because it’ll be baked into the code on the buidlserver. Instead we need to extend Blockly.ReplMgr.buildYail to send it. Alternatively, to cover both sending the form definition and “Do It”, you may want to change the behavior of Blockly.ReplMgr.putYail to always prepend the macro to the code.

**Comment by BeksOmega** [comment]
@ewpatton I got it working by modifying putYail :D works as expected for old versions and new versions.

[This](https://github.com/BeksOmega/appinventor-sources/pull/24#discussion_r479504701) is the last thing to resolve, and then we should be good.

File: appinventor/components/src/com/google/appinventor/components/runtime/OptionHelper.java (hunk 12 above)
Comment (ewpatton) [reply appinventor/components/src/com/google/appinventor/components/runtime/OptionHelper.java]: There are a number of situations where we might have a closed set of known options but for whatever reason the option set itself is open world. One example is ScreenOrientation, where the set of options available at SDK 7 or whatever is smaller than it is today. Granted we enforce this to be closed even though newer versions of Android might add more ScreenOrientation options. The second scenario is where a value might be returned by an external-to-App Inventor component. For example, in the case of the Web component we have a set of known HttpStatus codes (e.g., HttpStatus.OK, HttpStatus.NotFound). However, various applications built on top of HTTP might return status codes not defined in the original spec. So if we were to introduce an options item for HttpStatus, which we might use with Web's GotText event with its responseCode parameter, there's a reasonable set of options that make sense/are common to expose as an option but the set is technically open. A server could return error code 599, for example, which might be specific to the server's software. I think it's still valuable for us to allow an option list for common values but not keep the set closed.

**Comment by ewpatton** [comment]
@BeksOmega See my comment on the corresponding thread.

File: appinventor/components/src/com/google/appinventor/components/runtime/OptionHelper.java (hunk 12 above)
Comment (BeksOmega) [reply appinventor/components/src/com/google/appinventor/components/runtime/OptionHelper.java]: Gotcha! I'll add logic to make sure that if we get a concrete value which isn't defined in the OptionList we return the concrete value.
